File: tests/test_phase5.py
"""Phase 5: Flowsheet assembly, JSON serialisation, parser."""
import json
import math
import tempfile
from pathlib import Path
import pytest

from chemsim.flowsheet.flowsheet import Flowsheet
from chemsim.io.component_db import ComponentDB
from chemsim.thermo.peng_robinson import PengRobinson
from chemsim.thermo.flash import FlashCalculator
from chemsim.ops.mixer import MixerOp
from chemsim.ops.splitter import SplitterOp


DB_PATH = Path(__file__).resolve().parent.parent / "data" / "components.json"
EXAMPLES = Path(__file__).resolve().parent.parent / "examples"


def test_flowsheet_simple_recycle():
    fs = Flowsheet.from_json(
        str(EXAMPLES / "simple_recycle.json"),
        str(DB_PATH))
    ok = fs.solve()
    assert ok


def test_flowsheet_summary_contains_streams():
    fs = Flowsheet.from_json(
        str(EXAMPLES / "simple_recycle.json"),
        str(DB_PATH))
    fs.solve()
    summary = fs.summary()
    assert "FEED" in summary
    assert "ChemSim" in summary


def test_flowsheet_results_as_json():
    fs = Flowsheet.from_json(
        str(EXAMPLES / "simple_recycle.json"),
        str(DB_PATH))
    fs.solve()
    j = json.loads(fs.results_as_json())
    assert "streams" in j
    assert "components" in j
    assert "FEED" in j["streams"]


def test_flowsheet_export_results():
    fs = Flowsheet.from_json(
        str(EXAMPLES / "simple_recycle.json"),
        str(DB_PATH))
    fs.solve()
    with tempfile.NamedTemporaryFile(suffix=".json", delete=False) as f:
        path = f.name
    fs.export_results(path)
    with open(path) as f:
        data = json.load(f)
    assert "streams" in data


def test_flowsheet_distillation_recycle():
    fs = Flowsheet.from_json(
        str(EXAMPLES / "distillation_recycle.json"),
        str(DB_PATH))
    ok = fs.solve()
    assert ok
    dist = fs.get_stream("DISTILLATE")
    btms = fs.get_stream("BOTTOMS")
    assert dist.total_flow > 0.0
    assert btms.total_flow > 0.0


def test_flowsheet_programmatic_mixer_splitter():
    """Build a flowsheet programmatically: Feed → Mixer → Splitter → recycle."""
    db = ComponentDB(str(DB_PATH))
    comps = db.get(["METHANE", "ETHANE", "PROPANE", "N-BUTANE"])
    fs = Flowsheet(comps)

    fs.add_stream("FEED", T=260.0, P=2e6, flow=100.0,
                  composition={"METHANE": 0.4, "ETHANE": 0.3,
                               "PROPANE": 0.2, "N-BUTANE": 0.1})
    fs.add_stream("RECYCLE", T=260.0, P=2e6, flow=100.0,
                  composition={"METHANE": 0.4, "ETHANE": 0.3,
                               "PROPANE": 0.2, "N-BUTANE": 0.1})

    fc = fs.flash_calculator()
    fs.add_unit("MIX",   MixerOp(fc, ["fresh", "recycle"]))
    fs.add_unit("SPLIT", SplitterOp([0.2, 0.8]))

    fs.connect("FEED",    to_unit="MIX",   to_port="fresh")
    fs.connect("RECYCLE", from_unit="SPLIT", from_port="out1",
                          to_unit="MIX",   to_port="recycle")
    fs.connect("MIXOUT",  from_unit="MIX",   from_port="out",
                          to_unit="SPLIT", to_port="in")
    fs.connect("PRODUCT", from_unit="SPLIT", from_port="out0")

    ok = fs.solve()
    assert ok

    prod = fs.get_stream("PRODUCT")
    assert prod.total_flow > 0.0


def test_flowsheet_stream_names():
    fs = Flowsheet.from_json(
        str(EXAMPLES / "simple_recycle.json"),
        str(DB_PATH))
    names = fs.stream_names()
    assert "FEED" in names


def test_flowsheet_reset_to_base():
    fs = Flowsheet.from_json(
        str(EXAMPLES / "simple_recycle.json"),
        str(DB_PATH))
    fs.solve()
    T_before = fs.get_stream("FEED").T
    fs.set_stream_conditions("FEED", 300.0, 3e6)
    fs.reset_to_base()
    T_after = fs.get_stream("FEED").T
    assert abs(T_after - T_before) < 0.1


def test_flowsheet_results_table():
    fs = Flowsheet.from_json(
        str(EXAMPLES / "simple_recycle.json"),
        str(DB_PATH))
    fs.solve()
    table = fs.results_table()
    assert isinstance(table, list)
    assert len(table) > 0
    assert "total_flow" in table[0]

File: tests/test_phase5.cpp
#include <gtest/gtest.h>
#include <filesystem>
#include <fstream>
#include <sstream>
#include <nlohmann/json.hpp>
#include "chemsim/flowsheet/Flowsheet.hpp"
#include "chemsim/flowsheet/MixerOp.hpp"
#include "chemsim/flowsheet/SplitterOp.hpp"
#include "chemsim/io/ComponentDB.hpp"

#ifndef CHEMSIM_DATA_DIR
#define CHEMSIM_DATA_DIR "data"
#endif

using namespace chemsim;

static const std::string DB5 = std::string(CHEMSIM_DATA_DIR) + "/components.json";

namespace {

Flowsheet makeRecycleFlowsheet() {
    ComponentDB db(DB5);
    auto comps = db.get({"METHANE", "ETHANE", "PROPANE", "N-BUTANE"});
    Flowsheet fs(comps);

    const std::map<std::string, double> z = {
        {"METHANE", 0.40},
        {"ETHANE", 0.30},
        {"PROPANE", 0.20},
        {"N-BUTANE", 0.10}
    };

    fs.addStream("FEED", 260.0, 2e6, 100.0, z);
    fs.addStream("RECYCLE", 260.0, 2e6, 100.0, z);
    fs.addStream("MIXOUT", 260.0, 2e6, 100.0, z);

    fs.addUnit("MIX",
               std::make_unique<MixerOp>(fs.flashCalculator(),
                                         std::vector<std::string>{"fresh", "recycle"}));
    fs.addUnit("SPLIT",
               std::make_unique<SplitterOp>(std::vector<double>{0.2, 0.8}));

    fs.connect("FEED", "", "", "MIX", "fresh");
    fs.connect("RECYCLE", "SPLIT", "out1", "MIX", "recycle");
    fs.connect("MIXOUT", "MIX", "out", "SPLIT", "in");
    fs.connect("PRODUCT", "SPLIT", "out0", "", "");

    EXPECT_TRUE(fs.solve());
    return fs;
}

} // namespace

TEST(FlowsheetPolish, SummaryIncludesCoreStreamInfo) {
    auto fs = makeRecycleFlowsheet();
    const auto text = fs.summary();

    EXPECT_NE(text.find("ChemSim Flowsheet Summary"), std::string::npos);
    EXPECT_NE(text.find("PRODUCT"), std::string::npos);
    EXPECT_NE(text.find("RECYCLE"), std::string::npos);
    EXPECT_NE(text.find("phase="), std::string::npos);
}

TEST(FlowsheetPolish, PrintSummaryWritesToStream) {
    auto fs = makeRecycleFlowsheet();
    std::ostringstream out;
    fs.printSummary(out);
    EXPECT_EQ(out.str(), fs.summary());
}

TEST(FlowsheetPolish, ResultsJsonContainsSolvedStreams) {
    auto fs = makeRecycleFlowsheet();
    auto json = fs.resultsAsJson();

    ASSERT_TRUE(json.contains("streams"));
    ASSERT_TRUE(json["streams"].contains("PRODUCT"));
    EXPECT_NEAR(json["streams"]["PRODUCT"]["totalFlow"].get<double>(), 100.0, 1e-4);
    EXPECT_TRUE(json["streams"]["PRODUCT"].contains("phase"));
}

TEST(FlowsheetPolish, ExportResultsWritesJsonFile) {
    auto fs = makeRecycleFlowsheet();
    const auto path = std::filesystem::temp_directory_path() / "chemsim_results_phase5.json";

    fs.exportResults(path.string());

    std::ifstream in(path);
    ASSERT_TRUE(in.is_open());
    nlohmann::json json;
    in >> json;

    EXPECT_TRUE(json.contains("components"));
    EXPECT_TRUE(json.contains("streams"));
    EXPECT_TRUE(json["streams"].contains("FEED"));
    EXPECT_TRUE(json["streams"].contains("PRODUCT"));

    in.close();
    std::filesystem::remove(path);
}

TEST(FlowsheetPolish, UniqueJsonConstructionSolvesRecycleFlowsheet) {
    auto fs = Flowsheet::fromJSONUnique("examples/simple_recycle.json", DB5);

    ASSERT_TRUE(fs);
    EXPECT_TRUE(fs->solve());
    EXPECT_NEAR(fs->getStream("PRODUCT").totalFlow, 100.0, 1e-4);
    EXPECT_NEAR(fs->getStream("RECYCLE").totalFlow, 400.0, 1e-3);
}
